File: src/test/java/restfulbooker/api/ApiRequest.java
package restfulbooker.api;

import io.restassured.RestAssured;
import io.restassured.http.Method;

import io.restassured.response.Response;
import io.restassured.specification.RequestSpecification;

import java.util.HashMap;
import java.util.Map;

public class ApiRequest {
    private String url;
    private Method method;
    private Map<String, String> queryParams;
    private Map<String, String> headers = new HashMap<>();
    private Object body;

    public ApiRequest(Method method, String url) {
        this.method = method;
        this.url = url;
    }

    public ApiRequest withQueryParams(Map<String, String> queryParams) {
        this.queryParams = queryParams;
        return this;
    }

    public ApiRequest withBody(Object body) {
        if (body != null) {
            this.body = body;
        }
        return this;
    }

    public ApiRequest withHeaders(Map<String, String> headers) {
        if (headers != null) {
            this.headers.putAll(headers);
        }
        return this;
    }

    public Response sendRequest() {
        RequestSpecification request = RestAssured.given().log().all();
        if (headers != null && !headers.isEmpty()) {
            request.headers(headers);
        }
        if (queryParams != null &&!queryParams.isEmpty()) {
            request.queryParams(queryParams);
        }

        return switch (method) {
            case GET -> request.get(url).then().log().all().extract().response();
            case POST -> request.body(body).post(url).then().log().all().extract().response();
            case PUT -> request.body(body).put(url).then().log().all().extract().response();
            default -> throw new IllegalArgumentException("Invalid method type: " + method);
        };
    }
}

File: src/test/java/restfulbooker/api/ApiRequest.kt
package restfulbooker.api

import io.restassured.RestAssured
import io.restassured.http.Method
import io.restassured.response.Response
import io.restassured.specification.RequestSpecification

class ApiRequest(
    val method: Method = Method.GET, val url: String = "",
    val queryParams: Map<String, String?>? = null,
    val headers: Map<String, String?>? = null,
    val body: Any? = null,
) {

    fun sendRequest(): Response {
        val request: RequestSpecification = RestAssured.given().log().all()
        queryParams?.let {
            request.queryParams(it)
        }
        headers?.let {
            request.headers(it)
        }
        return when (method) {
            Method.GET -> request.get(url).then().log().all().extract().response()
            Method.POST -> request.body(body).post(url).then().log().all().extract().response()
            Method.PUT -> request.body(body).put(url).then().log().all().extract().response()
            Method.PATCH -> request.body(body).patch(url).then().log().all().extract().response()
            else -> throw IllegalArgumentException("Invalid method type: $method")
        }
    }


}